File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool checkBST(TreeNode* root, long long mn, long long mx) {
        if(!root) return true;
        return root->val > mn && root->val < mx && checkBST(root->left, mn, root->val) && checkBST(root->right, root->val, mx);
    }

    bool isValidBST(TreeNode* root) {
        return checkBST(root, LONG_MIN, LONG_MAX);
    }
};

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[]}
     */

    rightSideCall(node, level, ans) {
        if(!node) return;
        if(!ans[level]) {
            ans[level] = node.val
        }
        this.rightSideCall(node.right, level+1, ans);
        this.rightSideCall(node.left, level+1, ans);
    }

    rightSideView(root) {
        let ans = [];
        this.rightSideCall(root, 0, ans);
        return ans;
    }
}
